File: src/main/java/org/railwaystations/rsapi/core/model/Station.java
package org.railwaystations.rsapi.core.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.Value;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Value
@Builder
public class Station {

    private static final int EARTH_RADIUS = 6371;

    @Builder.Default
    Key key = new Key("", "0");

    String title;

    @Builder.Default
    Coordinates coordinates = new Coordinates(0.0, 0.0);

    String ds100;

    @Getter
    List<Photo> photos = new ArrayList<>();

    @Builder.Default
    boolean active = true;

    public boolean hasPhoto() {
        return this.photos.size() > 0;
    }

    public boolean appliesTo(String photographer) {
        if (photographer != null) {
            return hasPhoto() && photos.stream().anyMatch(photo -> photo.getPhotographer().getDisplayName().equals(photographer));
        }
        return true;
    }

    @Value
    @AllArgsConstructor
    public static class Key {
        String country;
        String id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Station other)) {
            return false;
        }
        return Objects.equals(key, other.getKey());
    }

    @Override
    public int hashCode() {
        return key.hashCode();
    }

}

File: core/src/main/kotlin/org/railwaystations/rsapi/core/model/Station.kt
package org.railwaystations.rsapi.core.model

data class Station(
    val key: Key = Key("", "0"),
    val title: String,
    val coordinates: Coordinates = Coordinates(),
    val ds100: String? = null,
    val photos: List<Photo> = emptyList(),
    val active: Boolean = true,
) {

    val hasPhoto: Boolean
        get() = photos.isNotEmpty()

    fun appliesTo(photographer: String?): Boolean {
        if (photographer != null) {
            return photos.any { it.photographer.displayName == photographer }
        }
        return true
    }

    val primaryPhoto: Photo?
        get() = photos.firstOrNull { it.primary }

    data class Key(
        var country: String,
        var id: String,
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is Station) {
            return false
        }
        return key == other.key
    }

    override fun hashCode(): Int {
        return key.hashCode()
    }

}
